### PR TITLE
Add image augmentation and a validation set split

### PR DESCRIPTION
**Pull request by Dexterp37**

### CODE REVIEW

File: profondo/cnn.py [hunk 1]
```diff
@@ -124,11 +135,30 @@ def report_callback(epoch, logs):
...
-              batch_size=args.batch_size,
-              epochs=args.epochs,
-              verbose=args.logging,
-              validation_data=(x_test, y_test),  # TODO: don't use test data
-              class_weight=genre_weights,
-              callbacks=callbacks)
+    if args.augment:
+        # Use ImageDataGenerator to augment images
+        datagen = ImageDataGenerator(
+            featurewise_center=False,
```
Comment (vitillo) [inline profondo/cnn.py]: I think it makes sense to do some centering here.

File: profondo/cnn.py [hunk 2]
```diff
@@ -124,11 +135,30 @@ def report_callback(epoch, logs):
...
+            featurewise_std_normalization=False,
+            rotation_range=90,
+            width_shift_range=0,
+            height_shift_range=0,
+            horizontal_flip=False,
```
Comment (vitillo) [inline profondo/cnn.py]: Flipping images might also make sense.

File: profondo/cnn.py [hunk 3]
```diff
@@ -124,11 +135,30 @@ def report_callback(epoch, logs):
...
-              verbose=args.logging,
-              validation_data=(x_test, y_test),  # TODO: don't use test data
-              class_weight=genre_weights,
-              callbacks=callbacks)
+    if args.augment:
+        # Use ImageDataGenerator to augment images
+        datagen = ImageDataGenerator(
+            featurewise_center=False,
+            featurewise_std_normalization=False,
+            rotation_range=90,
```
Comment (vitillo) [inline profondo/cnn.py]: I am not so sure about the rotation, does it really help?

File: profondo/cnn.py [hunk 4]
```diff
@@ -28,26 +29,35 @@ def load_data(path):
     return meta, images
 
 
-def train_test_split(meta, images, train_ratio=0.8):
+def train_test_split(meta, images, train_ratio=0.8, validation_ratio=0.2):
     idx = np.random.permutation(np.arange(len(images)))
```
Comment (vitillo) [inline profondo/cnn.py]: Let's set the seed for the random number generator so that if the training is re-run we keep using the same training data.

File: profondo/cnn.py [hunk 5]
```diff
@@ -28,26 +29,35 @@ def load_data(path):
...
     train_size = int(len(idx)*train_ratio)
-    train_idx = idx[:train_size]
-    test_idx = idx[train_size:]
+    validation_size = int(train_size*validation_ratio)
```
Comment (vitillo) [inline profondo/cnn.py]: I think the train ratio should be kept as specified. The user can then specify the train and validation ratio. The test ratio will be `1 - train_ratio - validation_ratio`.

**Review by vitillo — CHANGES_REQUESTED** [review]

File: profondo/cnn.py [hunk 6]
```diff
@@ -124,11 +136,27 @@ def report_callback(epoch, logs):
...
+        # Use ImageDataGenerator to augment images
+        datagen = ImageDataGenerator(
+            featurewise_center=True,
+            featurewise_std_normalization=False,
+            horizontal_flip=True,
+            vertical_flip=True)
+        datagen.fit(x_train)
+
+        model.fit_generator(datagen.flow(x_train, y_train, batch_size=args.batch_size),
+                  steps_per_epoch=2*args.batch_size,
```
Comment (vitillo) [inline profondo/cnn.py]: Why `2*args.batch_size`?

**Review by vitillo — CHANGES_REQUESTED** [review]

File: profondo/cnn.py (hunk 6 above)
Comment (Dexterp37) [reply profondo/cnn.py]: My mistake, this should be: "It should typically be equal to the number of unique samples of your dataset divided by the batch size." from [here](https://keras.io/models/sequential/)

**Review by vitillo — APPROVED** [review]
